### DISCUSSION TITLE
watermark removal by figuring out original color and opacity

### DISCUSSION THREAD

**Discussion by mishav78** [ImageMagick/ImageMagick#3602] [Digital Image Processing]
I'm trying to figure out what is the original color of the gray watermark and it's opacity when it was overlaid on top? Any ideas?
![Starbuck](https://user-images.githubusercontent.com/79035778/116170010-ef6adb80-a6d3-11eb-8b75-84b31c2807cf.jpg)

**Comment by fmw42** [comment]
One thought would be that the color is whatever it is on black background and the opacity would be determined from that color and whatever it is on the white background. Perhaps over-simplistic, but maybe a reasonable first approximation.

**Comment by snibgo** [comment]
Look carefully. There are at least three values of alpha, or two alphas and multiple shades of gray, or whatever.

The image is noisy, with heavy JPEG atifacts. You should use a lossless-compressed version, if possible.

From one image, I don't know an automatic method for finding a watermark. However, for this image, a semi-manual technique is simple, if we assume (pretend) that the gray patches have constant gray, and a constant alpha in the central portions.

A method is: 

1. Crop out one pixel that is the background near-black, and one pixel that is the background near-black plus gray mark.

2. Crop out one pixel that is white, and one pixel that is the white plus gray mark.

3. Now you have four images. Use the technique I show at http://snibgo.com/im/watermark.htm#twoimg . This will give a value for the gray shade and for the most-opaque alpha.

**Reply by mishav78** [reply]
I understand the part where you get the four pixels. Then you plug it into the formula that you lay out, but I don't understand what units is R and D in? Is alpha from 0.0 to 1.0 translated later to percentages from 0% to 100%. And then for color I don't know how to get from 0.0 to 1.0 to rgb at all unless you do the equation for red, then separately for green and then separately for blue. Can you please explain Snibgo. It says on that page if you know the opacity of a pixel you can calculate its color. I am not sure exactly how?

**Comment by snibgo** [comment]
What version of IM do you use, on what platform?

On that linked page, the variables S, Sa, R1, D1, R2 and D2 are in the range 0.0 to 1.0. In an image, this represents values between 0 and QuantumRange. For images with 8 bits per channel, QuantumRange is 255. For 16 bits per channel, QuantumRange is 65535.

On that link, I show a Windows BAT script findWater2img.bat that processes the four images and calculates the watermark that would make two of the images from the other two. If the inputs were one pixel, then the output calculated watermark will also be a single pixel. The colour and alpha of that pixel is what you want.

The script is essentially just one IM command. In you use bash, you will need to translate the script: change every caret ^ to backslash \ and every double %% to a single %.

The four images need to be same sizes. With this noisy image, the output watermark will depend heavily on what four inputs you choose. It may be better to use larger sizes to average-out the noise. For example, a crop of 24x3+436+733 is a gray-over-white area.

So we can crop the four images: white watermarked, white only, near-black watermarked, and near-black only.
```
magick ^
  nurseCoffee.jpg ^
  ( +clone -crop 24x3+436+733 +repage -write ww1.png +delete ) ^
  ( +clone -crop 24x3+362+810 +repage -write wo1.png +delete ) ^
  ( +clone -crop 24x3+690+221 +repage -write kw2.png +delete ) ^
  ( +clone -crop 24x3+690+203 +repage -write ko2.png +delete ) ^
  NULL:
```
Then use the script to calculate the watermark which we call ncw.png:
```
call %PICTBAT%findWater2img.bat wo1.png ko2.png ww1.png kw2.png ncw.png
```
Then find the average of ncw.png:
```
magick ^
  ncw.png ^
  -scale "1x1^!" ^
  txt:
```
The result is:
```
srgba(50.1882%,50.295%,50.2279%,0.502263)
```
Note that the RGB channels are not equal, so the result is not quite gray. We might repeat the processing for other locations in the image, including green patches, to increase the precision. But the result watermark is approximately: gray 50%, with 50% alpha. The watermark seems to be anti-aliased, with varying (reduced) alpha at the edges.

**Reply by mishav78** [reply]
7.0.10-62 On Windows 

Thanks for that answer Snibgo. So that beckons the second question which is even more important. 

Knowing the watermark is approximately #808080 and 50% opacity originally and even though it was antialiased causing noise how can we revert to approximately the original color that was present before the image was covered with a watermark. Please show the calculation for any one pixel. I assume the calculation will be the same for every other pixel white, black and green.

**Comment by snibgo** [comment]
When you have a watermark image and a watermarked-image, recovering the original input file is easy. See the "De-watermark" section of the link I showed.

Perhaps you can create a watermark image by drawing gray(50%) lines on a transparent background. Use the de-watermark method on your JPEG to get an approximation of the original. It won't be exact because your JPEG is noisy.

**Comment by snibgo** [comment]
Suppose you don't want the bother of drawing gray(50%) lines to try and re-create the watermark image, and that you are happy to have a very rough version of the original image. If so, you can use this method.

We make a guess at the watermark. We suppose the watermark image is the same size as your JPEG, and every pixel is gray(50%) at 50% opacity.
```
%IMG7%magick ^
  nurseCoffee.jpg ^
  -fill gray(50%%) -colorize 100 ^
  -alpha set ^
  -channel A -evaluate set 50%% +channel ^
  +depth ^
  PNG64:gw.png
```
![gw](https://user-images.githubusercontent.com/33812368/116331315-6b921b80-a7c7-11eb-8137-80a36c5de1f6.png)

Using this assumed watermark, recover the original image. In the command below, this is "dewm.miff". Where the pixels in this recovered image are outside the range zero to 100% of QuantumRange, no recovery is possible. To find these, we make a clone, clamp it to 0 to 100%, and find the difference. The result is black where the clamping didn't change the pixels.
```
%IMG7%magick ^
  nurseCoffee.jpg ^
  -define compose:clamp=off ^
  ( gw.png ^
    ( +clone -alpha extract +write mpr:WEX ) ^
    -alpha off ^
    -compose Multiply -composite ^
  ) ^
  -gravity center ^
  -compose MinusSrc -composite ^
  ( mpr:WEX -negate ) ^
  -alpha off ^
  -compose DivideSrc -composite ^
  -define quantum:format=floating-point -depth 32 ^
  +write dewm.miff ^
  ( +clone -clamp ) ^
  -compose Difference -composite ^
  -fill White +opaque Black ^
  +depth ^
  diff.png
```
diff.png is black where the assumed watermark worked, and white otherwise. It is noisy.
![diff](https://user-images.githubusercontent.com/33812368/116331353-7f3d8200-a7c7-11eb-95b2-653c82c54e51.png)

We use diff.png as a mask. Where is is white, we show pixels from your watermarked JPEG, and where it is black we show pixels from dewm.miff.
```
%IMG7%magick ^
  dewm.miff ^
  nurseCoffee.jpg ^
  diff.png ^
  -alpha off ^
  -compose Over -composite ^
  dewm2.png
```
The result, dewm2.png, is a very rough de-watermarked image.
![dewm2](https://user-images.githubusercontent.com/33812368/116331419-a431f500-a7c7-11eb-8d5f-abd079a4d6de.png)

**Reply by mishav78** [reply]
Wow, thanks for that Snibgo. And if I had several different images with the watermark in the center,  can the watermark be found automatically, even if the images are a different size? I was thinking cropping from the center with the gravity center and extent attributes.  
for example 
![image](https://user-images.githubusercontent.com/79035778/116377878-9e480e00-a7df-11eb-8f0a-6d356f7c57da.png)
and 
![image](https://user-images.githubusercontent.com/79035778/116378575-452caa00-a7e0-11eb-9cbe-ca5536da5941.png)
and 
![image](https://user-images.githubusercontent.com/79035778/116379279-dd2a9380-a7e0-11eb-941c-d3e832ca5445.png)

**Comment by snibgo** [comment]
On my watermark page, we assume the watermark image is the same size as the input image. We assume the watermarked image is made by:
```
magick {input-image} {watermark-image} -compose {method} -composite {watermarked-image}
```
... where the {method} is often "Over" but can be "Plus" or some other method.

If you want to de-watermark a number of different-sized images where the important parts of the watermark is the same, you will need to "-crop" or "-border" or "-extent" as required.

**Reply by mishav78** [reply]
so on that page you show how to find the watermark automatically?

**Comment by snibgo** [comment]
> so on that page you show how to find the watermark automatically?

I don't understand the question.

There are three images: the input image, a watermark image, and a watermarked image. We assume "-compose Over -composite" was used to make the watermarked image.

If we know the input and the watermark, we can calculate the watermarked image,

Or if we know the watermarked image and the watermark, we can calculate the input image.

But if we know the input image and the watermarked image, there are many possible watermarks that might have been used.

Alternatively, if we know two different inputs and the watermarked versions of those, we can calculate the watermark.

If we know only the watermarked image and nothing else, we cannot automatically calculate the watermark, or the input image. It is like asking: "What two numbers add up to one hundred?" There are many possible solutions, and we cannot know which is the correct one.

However, as I show in this thread, we can make assumptions: that each pixel in the watermark image is either transparent or an unknown constant alpha and an unknown constant colour, and that we know which pixels were transparent and which were that constant alpha. If we make those assumptions, we can calculate the alpha and the colour, so the watermark image is then known, and we can calculate the input image.

**Reply by mishav78** [reply]
I was thinking if you have twenty images with the same watermark and you stack them together each with five percent opacity as a composite . The watermark will appear the strongest with the rest noise. 
https://github.com/mishav78

**Comment by ashishb** [comment]
```
%IMG7%magick ^
  nurseCoffee.jpg ^
  -define compose:clamp=off ^
  ( gw.png ^
    ( +clone -alpha extract +write mpr:WEX ) ^
    -alpha off ^
    -compose Multiply -composite ^
  ) ^
  -gravity center ^
  -compose MinusSrc -composite ^
  ( mpr:WEX -negate ) ^
  -alpha off ^
  -compose DivideSrc -composite ^
  -define quantum:format=floating-point -depth 32 ^
  +write dewm.miff ^
  ( +clone -clamp ) ^
  -compose Difference -composite ^
  -fill White +opaque Black ^
  +depth ^
  diff.png
```

Dumb question but how do you run this script?

**Comment by snibgo** [comment]
It is Windows syntax. That command can be run either by pasting it into a command prompt, or pasting it into a file with ".bat" suffix and typing the name of that file in the command prompt.

There are other ways of running commands on Windows, which you can investigate if you are interested.

**Comment by ashishb** [comment]
Thanks. I'm failing to run them as a bash script.

Ashish

**Comment by fmw42** [comment]
In bash, replace ^ with \\  

replace   ( with \\(  

and replace   ) with \\). So

```
magick \
  nurseCoffee.jpg \
  -define compose:clamp=off \
  \( gw.png \
    \( +clone -alpha extract +write mpr:WEX \) \
    -alpha off \
    -compose Multiply -composite \
  \) \
  -gravity center \
  -compose MinusSrc -composite \
  \( mpr:WEX -negate \) \
  -alpha off \
  -compose DivideSrc -composite \
  -define quantum:format=floating-point -depth 32 \
  +write dewm.miff \
  \( +clone -clamp \) \
  -compose Difference -composite \
  -fill White +opaque Black \
  +depth \
  diff.png

```
